content/articles/lru-cache-implementation-in-javascript/index.md

```diff
@@ -0,0 +1,140 @@
+
+Another operation that can be performed on LRU is searching. Using our initial example, let us access the third item: red, from our cache.
+
+We first search to check whether `red` is available. Since it is available, it becomes the most recently accessed item. This means that we will have to shift `red` to the front of our cache or list. Our new cache order will be `[red, black, green, white]`.
+
+With LRU, we notice that when we access an item, it shifts to the front of the list. Thus, when we want to remove an item, we pick the least recently use item: the one at the end of the list, and remove it from our cache.
+
+### Data structures used to build an efficient LRU cache algorithm.
+From the example above, we can perform many operations while building the LRU cache. These operations include:
+
+#### 1. Searching
+When we want to get color `red` from our list, first, we need to check our entire list to check whether the item is there. 
+
+If the element is present, we reorder the list and move `red` to the beginning of the list. In this case, the time complexity for the re-ordering will be `O(n)`.
+
+#### 2. Adding items
+Let's say we have a scenario where we need to add e.g. black into our list. Our list, in this case, is still red, green, white, blue. Searching through the list, we can't find color black. The next thing we need to do since our cache size is 4, we remove blue from the list. Why remove blue? Because it is the least accessed item.
+
+Once that is done, we then shift red, green and white to the left. We are now left with an empty slot at the beginning of the list to add color black. The time complexity for doing the shifting is O(n).
+
+We can optimize the shifting operation by eliminating it. To do so we use a double liked list. How?
+
+Still using the example red, green, white and blue, we needed to add black. Since it wasn't in our list, what we did was remove the oldest element blue and shift the entire list items.
+
+In a doubly-linked list, all you need to do is to remove the rear item and update the rear element to the previous element. Then add a new node and make it point to the second element and update the front. The entire operation is 0(1) time complexity.
+
+This will be clearly seen in the code example below.
+
+We can also optimize searching. How?
+
+Still using our previous example of cache size 4, with `[red, green, white, blue]`. searching for black. We can optimize our search by using a hash table. With a Hash, you will have a key and a value pair. The Key in this case will be the actual values in our cache. The value will be the address of our value as shown in the diagram below.

```

Comment:
This whole section is still redundant. You just repeated what you said in the first section: What is LRU Cache?

 Please revisit this section removing all the redundancy.